### PR TITLE
feat(wal): segment file reference counting

### PR DESCRIPTION
**Pull request by domodwyer**
WAL file entries are wrote as data is ingested, and eventually that data is persisted. Once that data is persisted, it is no longer needed in the WAL file. When (and only when) all the entries the WAL file have been persisted, the file should be deleted.

This is tricky to do right, because of the lack of serialisation in the ingester (the core of our ingest performance). It leads to [hacks like this](https://github.com/influxdata/influxdb_iox/blob/bef29a683775aaa4b9e2cd66073c1115b64e4c44/ingester2/src/wal/rotate_task.rs#L37-L78) (which explains the problem in much more detail, and despite the danger danger, has a pretty low occurrence risk and impact).

This PR is part of a series to solve this timely WAL deletion problem. This PR implements a stateful, space-efficient tracker of (un)-persisted operations in each WAL segment file by tracking the IDs contained within those files, and counting how many unpersisted operations remain. Once there are no unpersisted operations, the file is safely deleted.

This code **is not yet used** but is PR-able on it's own. I'll follow up with metrics additions, and then plumbing it all in.

Part of https://github.com/influxdata/influxdb_iox/issues/6566.

---

* refactor(wal): const SegmentId constructor (d9ca8f948)
      
      Allow a SegmentId to be constructed in a const context.

* refactor: owned persisted SequenceNumberSet copy (f5c50094c)
      
      Allow an owned SequenceNumberSet to be destructured from an owned
      CompletedPersist notification.
      
      Additionally allow an owned SequenceNumberSet to be obtained from a
      (potentially) shared Arc<CompletedPersist> in a memory efficient / no
      cloning way. A CompletedPersist notification is typically wrapped in an
      Arc, and at time of writing, typically referenced only once.

* feat(wal): segement file reference counting (bef29a683)
      
      Reference count rotated WAL segment files, tracking the number of
      unpersisted operations in each and deleting the file once all the data
      within them has been persisted to object storage.
      
      This commit contains the implementation of the reference counting logic,
      and is currently unused. A follow-up PR will wire this into the various
      places needed to feed it the necessary information.
      
      Part of https://github.com/influxdata/influxdb_iox/issues/6566.

### CODE REVIEW

File: ingester2/src/wal/reference_tracker.rs [hunk 1]
```diff
@@ -0,0 +1,646 @@
...
+
+    /// Test in-order notifications:
+    ///
+    ///   * WAL file is rotated and the tracker notified
+    ///   * Multiple persists complete, and an unbuffered notification, draining
+    ///     the references to the file
+    ///   * The file is deleted when refs == 0
+    ///   * Dropping the handle stops the actor
+    #[tokio::test]
+    async fn test_rotate_persist_delete() {
```
Comment (domodwyer) [inline ingester2/src/wal/reference_tracker.rs:477]: This test outlines how this tracker will be utilised by the ingester.

File: ingester2/src/wal/reference_tracker.rs [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+    /// The returned [`WalReferenceActor`] SHOULD be
+    /// [`WalReferenceActor::run()`] before the handle is used to avoid
+    /// potential deadlocks.
+    pub(crate) fn new<T>(wal: T) -> (Self, WalReferenceActor<T>)
+    where
+        T: WalFileDeleter,
+    {
+        let (file_tx, file_rx) = mpsc::channel(5);
+        let (persist_tx, persist_rx) = mpsc::channel(50);
+        let (unbuffered_tx, unbuffered_rx) = mpsc::channel(50);
```
Comment (savage-engineer) [inline ingester2/src/wal/reference_tracker.rs:121]: Any reasoning behind these sizes out of interest?

File: ingester2/src/wal/reference_tracker.rs [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+
+/// A WAL file reference-count tracker.
+///
+/// See [`WalReferenceHandle`].
+#[derive(Debug)]
+pub(crate) struct WalReferenceActor<T = Arc<wal::Wal>> {
+    wal: T,
+
+    /// The set of IDs of persisted data that do not yet appear in
+    /// `wal_segments`, the set of WAL files rotated out of active use. This is
```
Comment (savage-engineer) [inline ingester2/src/wal/reference_tracker.rs:190]:
```suggestion
    /// `wal_files`, the set of WAL files rotated out of active use. This is
```
is this meant to be the variable you refer to here?

**Review by savage-engineer — APPROVED** [review]

File: ingester2/src/wal/reference_tracker.rs (hunk 2 above)
Comment (domodwyer) [reply ingester2/src/wal/reference_tracker.rs:121]: Complete guesstimates based on what I expect to be possible, plus a load of headroom. 

Element size is small so I went big.

**Pull request merged**